File: cpp/observer/observer.cpp
#include <iostream>
#include <stdio.h>
#include <vector>

using namespace std;

// an object is something being observed
class CObject;

// a simple object - an observer: someone, who should be
// notified when the something changes 
class CObserver {
public:
	virtual void onSubjectChanged(CObject *who) = 0;
	virtual ~CObserver() { }
};

// a simple object, that allows to be observed
class CObject {
private:
	// a private vector of current observers
	vector<CObserver*> _observers;
public:
	// a method to register an observer - that is, to say that a new observer 
	// now wants to be notified when the object changes
	void addObserver(CObserver *o) {
		_observers.push_back(o);
	}
	// a method that notifies all current observers
	void notifyEveryone() {
		for (const auto &o : _observers)
			o->onSubjectChanged(this);
	}
	virtual ~CObject()
	{ }
};

// an example of an object - an Animal in the Zoo,
class CAnimal : public CObject {
public:
	// last said phrase
	string lastsound;
	// animal name
	string name;
	// constructor
	CAnimal(string aname) {
		name = aname;
		lastsound = "";
	}
	// animal performs an action, and every observer is notified
	void say(string sound) {
		lastsound = sound;
		printf("\nCAnimal %s : '%s'\n", name.c_str(), sound.c_str());
		// notify every observer that is watching
		notifyEveryone();
	}
};

// an example of an observer - a child in the zoo, watching animals
class CChild : public CObserver {
public:
	// child name
	string name;
	// constructor
	CChild(string cname) {
		name = cname;
	}
	// a method which is called for a child when the animal changes,
	// the animal actually calls this method, passing itself as a parameter,
	// so the child not only knows THAT something has changed, but
	// also knows WHO was changed
	void onSubjectChanged(CObject *who) {
		printf("CChild %s : 'Hurray! %s had just said : %s'\n", name.c_str(), ((CAnimal*)who)->name.c_str(), ((CAnimal*)who)->lastsound.c_str());
	}
};

void main() {
	// create animals
	CAnimal cat("Zorra");
	CAnimal dog("Tortilla");
	CAnimal unknown("Cheburashka");
	
	// create children
	CChild vasya("Vasya");
	CChild petya("Petya");
	CChild kolya("Kolya");
	
	// assign who is watching (observing) who
	cat.addObserver(&vasya);
	cat.addObserver(&petya);
	dog.addObserver(&kolya);
	
	// check that when the animals change, children get notified;
	// the main idea of observer pattern is that objects can notify each
	// other indirectly, WITHOUT HARDCODING which objects exactly update which 
	cat.say("mew-mew!");
	dog.say("wow-wow!");
	cat.say("wow-wow!");
	dog.say("mew-mew!");
	unknown.say("nobody cares about me");

	system("pause"); 
}

File: csharp/observer/observer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Observer
{
	// наблюдатель - кто-то, кто хочет, чтобы его уведомили, 
	// когда что-то произойдет 
	abstract class CObserver
	{
		public abstract void onSubjectChanged(CSubject who);
	};

	// наблюдаемый объект - кто-то, кто позволяет за собой следить 
	class CSubject
	{
		// список текущих наблюдателей этого наблюдаемого объекта
		private List<CObserver> _observers;
		public CSubject()
		{
			// создаем пустой список наблюдателей - пока никто не наблюдает
			_observers = new List<CObserver>();
		}
		// метод для регистрации нового наблюдателя, с помощью которого наблюдатель
		// подписывается на уведомления о том, что наблюдаемый объект изменился
		public void addObserver(CObserver o)
		{
			_observers.Add(o);
		}
		// метод, который уведомляет всех подписанных наблюдателей о том, что что-то изменилось
		public void notifyEveryone()
		{
			// проходим по всем наблюдателям и вызываем у каждого его метод onSubjectChanged
			foreach (CObserver o in _observers)
				o.onSubjectChanged(this);
		}
	};

	// пример наблюдаемого объекта - животное в зоопарке
	class CAnimal : CSubject
	{
		// последняя сказанная фраза
		public string lastsound;
		// имя животного
		public string name;
		public CAnimal(string aname)
		{
			name = aname;
			lastsound = "";
		}
		// животное выполняет действие, и каждый наблюдатель получает уведомление
		public void say(string sound)
		{
			lastsound = sound;
			Console.WriteLine(string.Format("\nCAnimal {0} : '{1}'", name, sound));
			// вызываем уведомление каждого наблюдателя
			notifyEveryone();
		}
	};

	// пример наблюдателя - ребенок в зоопарке, наблюдающий за животными
	class CChild : CObserver
	{
		// имя ребенка
		public string name;
		public CChild(string cname)
		{
			name = cname;
		}
		// метод ребенка, вызываемый тогда, когда наблюдаемый им объект изменяется, 
		// животное вызывает этот метод, передавая самого себя как параметр,
		// поэтому ребенок не только получает уведомление о том, что кто-то изменился,
		// но и знает, кто изменился
		public override void onSubjectChanged(CSubject who)
		{
			Console.WriteLine(string.Format("CChild {0} : 'Hurray! {1} had just said : {2}'", name, 
					 (who as CAnimal).name, (who as CAnimal).lastsound));
		}
	};

	class observer
	{
		static void Main(string[] args)
		{
			// создаем животных
			CAnimal cat = new CAnimal("Zorra");
			CAnimal dog = new CAnimal("Tortilla");
			CAnimal unknown = new CAnimal("Cheburashka");
	
			// создаем детей
			CChild vasya = new CChild("Vasya");
			CChild petya = new CChild("Petya");
			CChild kolya = new CChild("Kolya");
	
			// прописываем, кто наблюдает за кем
			cat.addObserver(vasya);
			cat.addObserver(petya);
			dog.addObserver(kolya);
	
			// заметьте, когда животное изменяется, его наблюдатели уведомляются, 
			// таким образом, объекты могут уведомлять друг друга, причем эта связь
			// может возникать во время работы программы, а не прописываться жестко
			// до запуска программы
			cat.say("mew-mew!");
			dog.say("wow-wow!");
			cat.say("wow-wow!");
			dog.say("mew-mew!");
			unknown.say("nobody cares about me");

			Console.ReadKey();
		}
	}
}
